Add ability to pass customMetadata to google structured data ld_json tag

B77Mills opened the pull request:
In the case of FCP they are adding  articleSection: primarySection, creators: [author.name] & keywords: [taxonomy.name].

**FCP Company:**
<img width="877" alt="Screen Shot 2021-06-28 at 11 30 07 AM" src="https://user-images.githubusercontent.com/3845869/123671787-455b1d00-d804-11eb-9242-403f0d67d8cf.png">
**FCP Article**
<img width="1058" alt="Screen Shot 2021-06-28 at 11 30 26 AM" src="https://user-images.githubusercontent.com/3845869/123671790-45f3b380-d804-11eb-8045-52883a17d6a9.png">
**Standard Article**
<img width="1067" alt="Screen Shot 2021-06-28 at 11 35 02 AM" src="https://user-images.githubusercontent.com/3845869/123672457-f19d0380-d804-11eb-9f63-f4532b8dc69d.png">

B77Mills committed:
Add Ability to pas in custom data

Add ability to pass in customMetadata object to spread onto preset google structured data.

B77Mills committed:
add required fields to pass custom Metadata

Add need calls to pass customMetada from page layout to generation of metadata.

B77Mills committed:
missed the spreading it on the thrid match

B77Mills committed:
remove custon data object and pass builder and fragment

Update marko web to take a builder function and a fragment in order to generate the ld+json script dependent on site/org requiremnts.

zarathustra323 commented on packages/marko-web/components/page/metadata/content.marko:
```diff
@@ -1,12 +1,12 @@
 
 $ const { id } = input;
-$ const queryFragment = gql`
+$ const defaultQueryFragment = gql`
```
Since the fragment defined here is required for this component to work, it would be be better to "spread" the custom fragment into the existing one.... this will prevent issues if the incoming fragment lacks any of the required fields needed by this component

solocommand commented on packages/marko-web/components/page/metadata/content.marko:
```diff
@@ -52,6 +52,9 @@ fragment ContentPageMetadataFragment on Content {
 `;
 
+$ const queryFragment = input.structuredDataQueryFragment || defaultQueryFragment;
+$ const buildStructuredData = input.buildStructuredData || defaultBuildStructuredData;
```
Use the `@parameter1/base-cms-utils` `isFunction` utility here to ensure that the input argument is actually executable.

solocommand commented on packages/marko-web/components/page/layouts/marko.json:
```diff
@@ -38,6 +38,8 @@
     "<above-page>": {},
     "<below-page>": {},
+    "@structured-data-query-fragment": "string",
+    "@build-structured-data": "string",
```
This shouldn't be a string :)

solocommand commented on packages/marko-web/components/page/layouts/content.marko:
```diff
@@ -7,7 +7,7 @@ $ const { document } = out.global;
 >
   <@head>
-    <marko-web-content-page-metadata id=id />
+    <marko-web-content-page-metadata id=id structured-data-query-fragment=structuredDataQueryFragment build-structured-data=buildStructuredData />
```
Line length, break these params into new lines por favor

solocommand commented on packages/marko-web/components/page/metadata/marko.json:
```diff
@@ -8,7 +8,9 @@
-    "@id": "number"
+    "@id": "number",
+    "@structured-data-query-fragment": "string",
+    "@build-structured-data": "string"
```
Same here, not a string

B77Mills committed:
update structured date for all content types

@todo:

determin the proper use of
```
mainEntityOfPage: {
        '@type': 'WebPage',
        '@id': get(node, 'siteContext.canonicalUrl'),
      },
```
vs
```
url: canonicalUrl,
    ...(siteUrl !== canonicalUrl && { url: siteUrl, isBasedOn: canonicalUrl }),
```

B77Mills committed:
Spread cutom fragment on to queryFragment for cutom data

when custom field fragment is passed spread it onto the default queryFragment instead of overriding it.  Prevent removal of required fields.

B77Mills committed:
Add check to ensure buildStructureData isFunction

B77Mills committed:
fixing input type and readability

solocommand commented:
@B77Mills is this still needed?

B77Mills committed:
add missing util import for isFunction

B77Mills commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Parse.ly code bug - explanation included.](https://trello.com/c/kTZSYFUU/100-parsely-code-bug-explanation-included)

zarathustra323 commented on packages/marko-web/components/page/metadata/google-structured-data/content.js:
```diff
@@ -20,58 +20,89 @@ const getImages = (node) => {
-    const structuredData = JSON.stringify({
-      '@context': 'https://schema.org',
-      '@type': 'Article',
-      mainEntityOfPage: {
```
Move `mainEntityOfPage` to default structured data

B77Mills committed:
add mainEntityOfPage to defaultStructuredData

zarathustra323 commented on packages/marko-web/components/page/metadata/google-structured-data/content.js:
```diff
@@ -20,58 +20,93 @@ const getImages = (node) => {
   }
 
-  return undefined;
+  return JSON.stringify(defaultStruturedData);
```
I still think we need to return undefined so we're not assuming structured data for other content types

B77Mills committed:
Return undefined instead of defaultStructureData
